staging/src/k8s.io/kms/encryption/service.go

```diff
@@ -219,12 +273,141 @@ func (m *LocalKEKService) Decrypt(ctx context.Context, uid string, req *service.
+			// TODO(aramase): we are currently not returning the error if rotation fails. We should
+			// allow the failed state for an arbitrary time period and return the error if the state
+			// is not eventually fixed.
+		}
+	}
+
+	var aggregateHealthz []string
+	if r.Healthz != "ok" {
+		aggregateHealthz = append(aggregateHealthz, r.Healthz)
+	}
+
+	if !m.isReady.Load() {
+		// if the localKEKService is not ready, we will set the healthz status to not ready
+		klog.V(2).InfoS("localKEKService is not ready", "keyID", r.KeyID)
+		aggregateHealthz = append(aggregateHealthz, "localKEKService is not ready")
+	}
+	if len(aggregateHealthz) > 0 {
+		r.Healthz = strings.Join(aggregateHealthz, "; ")
+	}
+
+	return r, nil
 }
 
+// rotateLocalKEK rotates the local KEK by generating a new local KEK and encrypting it with the
+// remote KMS.
+func (m *LocalKEKService) rotateLocalKEK(ctx context.Context, expectedKeyID string) error {
+	uid := string(uuid.NewUUID())
+	if err := m.generateLocalKEK(ctx, uid, expectedKeyID); err != nil {
+		klog.V(2).ErrorS(err, "failed to generate local KEK as part of rotation", "uid", uid)
+		return fmt.Errorf("[uid=%s] failed to generate local KEK as part of rotation: %w", uid, err)
+	}
+	return nil
+}
+
+// generateLocalKEK generates a new local KEK and encrypts it with the remote KMS.
+// if expectedKeyID is not empty, it will check if the keyID returned from the remote KMS matches
+// the expected keyID. If the keyID does not match, it will continue using the existing local KEK
+// and return an error.
+func (m *LocalKEKService) generateLocalKEK(ctx context.Context, uid, expectedKeyID string) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	lk := m.getLocalKEK()
+	// if the localKEK was generated in the last pollInterval duration, we will not generate a new
+	// localKEK. This is to avoid regenerating a new localKEK for queued requests.
+	if lk.transformer != nil && m.clock.Since(lk.generatedAt) < m.pollInterval {
+		return nil
+	}
+
+	key, err := generateKey(keyLength)
+	if err != nil {
+		return fmt.Errorf("failed to generate local KEK: %w", err)
+	}
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return fmt.Errorf("failed to create cipher block: %w", err)
+	}
+
+	resp, err := m.remoteKMS.Encrypt(ctx, uid, key)
+	if err != nil {
+		return fmt.Errorf("failed to encrypt local KEK: %w", err)
+	}
+	if err = validateRemoteKMSEncryptResponse(resp); err != nil {
+		return fmt.Errorf("invalid response from remote KMS: %w", err)
+	}
+	if expectedKeyID != "" && resp.KeyID != expectedKeyID {
+		return fmt.Errorf("keyID returned from remote KMS does not match expected keyID")
+	}
+
+	m.localKEKTracker.Store(&localKEK{
+		encKEK:            resp.Ciphertext,
+		expiry:            m.clock.Now().Add(m.keyMaxAge),
+		usage:             atomic.Uint64{},
+		transformer:       aestransformer.NewGCMTransformer(block),
+		remoteKMSResponse: copyResponseAndAddLocalKEKAnnotation(resp),
+		generatedAt:       m.clock.Now(),
```

Comment:
Lets use the same `now` value here to be consistent in what we store:

```suggestion

	now := m.clock.Now()
	m.localKEKTracker.Store(&localKEK{
		encKEK:            resp.Ciphertext,
		expiry:            now.Add(m.keyMaxAge),
		usage:             atomic.Uint64{},
		transformer:       aestransformer.NewGCMTransformer(block),
		remoteKMSResponse: copyResponseAndAddLocalKEKAnnotation(resp),
		generatedAt:       now,
```